### PR TITLE
Support parse_url function

### PR DESCRIPTION
**Pull request by jackylee-ch**
support parse_url in gandiva

### CODE REVIEW

File: cpp/src/gandiva/function_holder_registry.h [hunk 1]
```diff
@@ -17,70 +17,129 @@
...
+            memcpy(result_buffer, out.data(), *out_length);
+
+            return result_buffer;
+        }
+    }
+
+    private:
+    explicit ParseUrlHolder(const std::string& extract_part) : extract_part_(extract_part) {}
+
+    std::string extract_part_;
```
Comment (zhouyuan) [inline cpp/src/gandiva/function_holder_registry.h:143]: move these to "parse_url_holder.h" ?

File: cpp/src/gandiva/parse_url_holder.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/util/uri.h"
```
Comment (zhouyuan) [inline cpp/src/gandiva/parse_url_holder.h:20]: it looks like some headers are missing
#include "arrow/status.h"
#include "gandiva/function_holder.h"
#include "gandiva/node.h"
#include "gandiva/visibility.h"

**Review by zhouyuan — COMMENTED** [review]
looks in a good shape!

File: cpp/src/gandiva/parse_url_holder.h (hunk 2 above)
Comment (jackylee-ch) [reply cpp/src/gandiva/parse_url_holder.h:20]: Yes, I also miss this `#include <unordered_map>`.

File: cpp/src/gandiva/function_holder_registry.h (hunk 1 above)
Comment (jackylee-ch) [reply cpp/src/gandiva/function_holder_registry.h]: yeap,wrong code.

File: cpp/src/gandiva/parse_url_holder.h [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        if (extract_part_ != "QUERY" || !uri.has_query()) {
+            return nullptr;
+        } else {
+            std::unordered_map<std::string, std::string> queries;
+            const auto items = std::move(uri.query_items()).ValueUnsafe();
+            for (const auto& query : items) {
+                queries.emplace(query.first, query.second);
+            }
+
+            auto out_query = queries.find(pattern_string);
```
Comment (PHILO-HE) [inline cpp/src/gandiva/parse_url_holder.h:117]: It looks like pattern is viewed as a regex in the implementation of vanilla spark. Please check this part further.
If needed, you can use `re2` lib to find the matched part. Please refer to `extract_holder.h` to see how to call the APIs in `re2`.

File: cpp/src/gandiva/parse_url_holder.cc [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        ARROW_RETURN_IF(
+                !IsArrowStringLiteral(extract_part_type),
+                Status::Invalid(
+                        "'parse_url' function requires a string literal as the second parameter"));
+
+        if (node.children().size() == 3) {
+            auto pattern = dynamic_cast<LiteralNode*>(node.children().at(2).get());
+            ARROW_RETURN_IF(
+                    pattern == nullptr,
+                    Status::Invalid("'parse_url' function requires a literal as the third parameter"));
```
Comment (PHILO-HE) [inline cpp/src/gandiva/parse_url_holder.cc:46]: Can we support non-literal case? If literal is enough in the workload, please also add the logic for checking literal in scala code and let the projection fall back for non-literal cases in `buildCheck`.

File: cpp/src/gandiva/parse_url_holder.h (hunk 3 above)
Comment (jackylee-ch) [reply cpp/src/gandiva/parse_url_holder.h:117]: It seems that spark uses regexp to obtain query data, and there is no regular-related case in the test case. I'm not sure if it's necessary to support regex here.

File: cpp/src/gandiva/parse_url_holder.cc (hunk 4 above)
Comment (jackylee-ch) [reply cpp/src/gandiva/parse_url_holder.cc:46]: It is a good question.After my local test is passed, I will consider supporting non-literal cases.

File: cpp/src/gandiva/parse_url_holder.h (hunk 3 above)
Comment (PHILO-HE) [reply cpp/src/gandiva/parse_url_holder.h:117]: OK. We can keep the current impl to just support plain string in this PR. Please leave a comment to clarify this.

Comment (jackylee-ch) [reply cpp/src/gandiva/parse_url_holder.h:117]: Sure.

File: cpp/src/gandiva/parse_url_holder.cc (hunk 4 above)
Comment (jackylee-ch) [reply cpp/src/gandiva/parse_url_holder.cc]: I have changed the code to support non-literal cases.

File: cpp/src/gandiva/gdv_function_stubs.h [hunk 5]
```diff
@@ -69,8 +69,18 @@ const char* gdv_fn_regexp_extract_utf8_utf8_int32(
     int64_t ptr, int64_t holder_ptr, const char* data, int32_t data_len,
     const char* /*pattern*/, int32_t /*pattern_len*/, const int32_t idx, int32_t* out_length);
 
-const char* gdv_fn_substr_index_utf8_utf8_int32(int64_t ptr, int64_t holder_ptr, const char* input, int in_len,
-                                         const char* delim, int delim_len, int count, int32_t* out_len);
+const char* gdv_fn_parse_url_utf8_utf8(
```
Comment (zhouyuan) [inline cpp/src/gandiva/gdv_function_stubs.h:72]: this func def should be removed, otherwise the building failed
fixed #139

Comment (jackylee-ch) [reply cpp/src/gandiva/gdv_function_stubs.h:72]: Thanks for your fix.

**Pull request merged**